### ISSUE TITLE
Objects that become invalid panic the renderer

### ISSUE THREAD

**Issue by danderson** [digitalrebar/provision#873]
I have a workflow that goes through install-ubuntu-18.04, reboots to local, then runs install-docker and krib-install.

When I apply that workflow to a machine, it gets all the way to krib-install, then stalls. The terminal in which I'm running DRP logs a panic every few seconds - I'm assuming the runner on the machine is repeatedly asking what to do next, and that asking is what triggers the panic, so it stays wedged there indefinitely.

```
2018/05/14 01:50:12 [Recovery] panic recovered:
GET /api/v3/jobs/26e7e303-65bd-4974-9824-6b43bb1c0494/actions HTTP/1.1
Host: 192.168.17.66:8092
Accept: application/json
Accept-Encoding: gzip
Authorization: Bearer _kRsh0W5aBDzfBpLHx63VpQLdVAjVNpMdo2gedHaCJxTUtYqcJASAnwyNsohN-XzoItqyYgsdY15hr4wubSf0NiyRxnpNrl4N4FpxkcJhMYR6GoMgKJ4x9iAgIeMWbTfCC1Fpaeo71yli8MyMdF-F80rI0PkQCEoFr2M2yo-Xpq3L6RHzMHzW-xZ6Ab0JkgL6s_cGyPyILOcOlNI9kCazZGuBpIIKBrSvBwhi5Yzf5IKEJ7w00zgVvw99Oqwu-O40Zvv0Qh4YE93jmQosqr9xhyVuM5-rdaePFY3fL_UEKGknLicnqKapIKAHpBO2jP0VZl7DdT6_ExdL71RAXZPpWS6S7DNJqjue-mNIeEBxu1buxC9YPESmShqmN-wYxFHCftLbR1rKzOJlhwTuGeX_RIA4r_Ka9WbX2b5KP0rBFNwiFoinmX2nlq8d8m9geo4BuSsJZpKSJcWsGPsX3X2LpeZ_WNDWnjs95xo0LAIFfMNUaVRM_RtXN6Jbel3Y8zpRuP-YPW1gJABVvFe-0NESau_kLemVZVHnnyyiLM7ewybT7H7GoPNtkZgz9eCgh2OHus6roW7oGJhH4Yx0RVImObMjDWPpnwXJEyxBtterRTyZkDu2tEv8waqBfurnvCLVzBz0YAMrBP9lhirR2KK1o-1MGu3qApYw2vcvmhmonU98WivAE2DyE63S4kjGu2XXFTW6TIpGUn-9v5stz_5TVW55R1z-4drq2_RsalzO7I0M29hGztNaC-NX958pkSZYaknGsgEQv13jSob4CRt6kuLFG9h9MWkTmLjQDykxjhawwCHH3LvtR8ijcS0UClTbjfz3jodNNeiuCkNwA6ih8jQ2zAKFHaDoRIyCVebuzlWYboyP4DpbDQ-VS-PbQe1EBxZ2KxNvNlgpVuC6SpJbPkUKC-ifAb_dhYigoCJmygz2LtMyxvcK4y-0IF0k8MMwclpXArAyNOensSjMwGD15y-hGd3RfXcFUdZFZ2ZTpkzsbk0dZu2hicK2vodJy9DU-7EMaf7cwBmnN0CBaOY_f6TJJFCy4yrMPd0s0QSaOhkEgFHEIjYqZQcJR24CNqR-7pZsxCfsXABbsRu9s4ikf6Q-e3Jd7bv5tNwP5FKdc1jZRStxg532v-uu6OKnG0k1OWuPDUSxzMcoJTrdBw1YW7-suqCUSRUBYZmES3VnqSBrLMXniQjxwMukWqCQGN_hd_u0t-ySmlnneLoysvfRcqKRzoBscm-jwKzBK3GDrFyrGts_lbE4BPBGmRaGpgdbfKsM7Wa5rM5kHKAgLUP3U0AYbukjVmqFqrHmfjQErzTYhrmZtpvCe4QUeiIBK7-LHrsKiP3Ky3QPcxpIJlm9qq3rBslUJPs3sZhfyL_IreHv5Lw-VXB13xzxY7qjT1d7fRoZzZYp60wd_Nq2IHBgAl41VW1P6gr5Sa20c4ol_Xcx86BUu0RoaWQxMivcD12uYbsBvYPyPZhTilqGnfVN1DKDInhwYu0s7nOHdTKCtG10KmnBnN35MSYphRCqyWf0oDRlPEu3IjXa0tOGFmWmE01rdjQdaZxqyYn7mHpi6Ltd4KxdBqrna05cvjpETlxtUIvnMCAy8mbB_uGrqDI_45MDoBDZxTbaE4kjYymBvjHEuOgLwHKYvvPVLbv2GfuSM2iNX2OYIxZeoRcFi5y_JtBMPT4OXelRgoT7hAfxBj-Q3AJ3qy42lv6U6FhTYopzXXMpkSbBr6xZ9OYW4OnIWQ76DkErrhI6vEDg0WMVa2rQjeqZB74FazG-qedDDxhE5g01mqOCPRrj_gHmn0MGZY6nh93Ocu6sg_xNmvxJ8f7d7A7IWhK5s3J0gjB7LhuL0c5ja6nKWJsCfFe5vSh-7y9PlCB0XEzAApFMJe1POCbpz83NRkAPkrmBnmG8NKUkMNac9TSsPpbGfduuu-ND2WszAlp3IiYEhvT9YAUiDUHDUh697hMfZmSfnv-ADrxIECutKngrbGJa6_jJBktaBKUvKYg84D6HYpyQHvcHzKWcAMHnE-CQEV8d6Es6gumS7CO6aT0B7vfyST9ImWO2w3iBYMiJb_3PPfSoAQ1FuarfjJHq13GmPTalGbL7_oVXanf-RdpECJonrIyu_W-Z5_58PCbpgo1wxFCOwdGmGQ1Y3HBtz-PyNJkkVLQZDHqpxHP16RDsz5b7NC6C1dXVynu2eU-WS5KDkORAJmDwpzFhpXloVfsH4rgWip5r2eyPSxbs4hBP-YcVwhMgSugidndpAj4JJyINdaA3YpuYyCcN9btYXDzh-IVbuGyaVIeWf8mPjYRQOu62njubRJXUFONT-whIfCUjT10qGD0etD_pT1iSHmft4FwD_olR3xf9oegpR7miaCZACNJhd35eL-fyRjr
User-Agent: Go-http-client/1.1


runtime error: invalid memory address or nil pointer dereference
/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/panic.go:491 (0x42c762)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/panic.go:63 (0x42b66d)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/signal_unix.go:367 (0x442f9b)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/text/template/template.go:179 (0x90aadc)
/home/travis/gopath/src/github.com/digitalrebar/provision/backend/renderData.go:147 (0x90aadb)
/home/travis/gopath/src/github.com/digitalrebar/provision/backend/jobs.go:553 (0x8c63ca)
/home/travis/gopath/src/github.com/digitalrebar/provision/frontend/jobs.go:629 (0xbc9f49)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/context.go:107 (0xb27082)
/home/travis/gopath/src/github.com/digitalrebar/provision/frontend/frontend.go:371 (0xbbe6cc)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/context.go:107 (0xb27082)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/recovery.go:46 (0xb38439)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/context.go:107 (0xb27082)
/home/travis/gopath/src/github.com/digitalrebar/provision/frontend/frontend.go:430 (0xbbfa0f)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/context.go:107 (0xb27082)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/gin.go:359 (0xb30065)
/home/travis/gopath/src/github.com/digitalrebar/provision/vendor/github.com/gin-gonic/gin/gin.go:326 (0xb2f7fa)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2619 (0x6b4013)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:1801 (0x6b015c)
/home/travis/.gimme/versions/go1.9.linux.amd64/src/runtime/asm_amd64.s:2337 (0x45b720)
```

I'm going to dig through the code to see if I'm holding it wrong (likely), but DRP should probably still not panic when whatever it is happens.

Since this is a testbed universe I'm playing in, I'm happy to provide a copy of drp-data/ if it'd help debug.

**Comment by galthaus** [comment]
@danderson - Did you add a profile to all the krib machines for cluster management?  And set the cluster-name parameter in it?

Also the jobs section of the UI or the machine task list should show which task is specifically failing or the last one to succeed. This would help with which part of the krib-install stage is having issues.

**Comment by danderson** [comment]
Yup, I have a `k8s-prod` profile, whose only param is `krib/cluster-profile: "k8s-prod"`. That profile is applied to a machine.

Looking at job logs, the spinning job is for the krib-install task. The job log repeats the following lines in a loop:

```
Starting task krib-install on 96f97f3a-51a0-4731-bb09-7e7439eca940
Failed to render actions: CLIENT_ERROR
  Cannot handle content-type text/plain; charset=utf-8
  No decoder for content-type text/plain; charset=utf-8
Updated job 26e7e303-65bd-4974-9824-6b43bb1c0494 to incomplete
```

**Comment by danderson** [comment]
Oh, and now I feel dumb, you said cluster-name, not cluster-profile. I'll go do that...

**Comment by danderson** [comment]
Nope, still causes panics. The krib profile is now:

```
$ ./drpcli profiles show k8s-prod
{
  "Available": true,
  "Description": "Production Kubernetes cluster",
  "Errors": [],
  "Meta": {
    "color": "red",
    "icon": "sitemap",
    "title": "Production k8s"
  },
  "Name": "k8s-prod",
  "Params": {
    "krib/cluster-name": "k8s-prod",
    "krib/cluster-profile": "k8s-prod"
  },
  "ReadOnly": false,
  "Validated": true
}
```

The failure mode is unchanged from earlier.

I'm using 3.8.2, is it worth me trying tip to see if that fares any better?

**Comment by danderson** [comment]
And for good measure, I set up another workflow for an in-memory cluster on Sledgehammer, in case the base ubuntu was the issue. Fails exactly the same, panic on getting actions for the krib-install task.

**Comment by danderson** [comment]
Since I'm bored and curious to learn about the source code structure, I did a bit of digging. Hopefully this'll be helpful somehow.

The crash occurs at `backend/renderData.go:147`, which is `tmpl := rd.target.templates().Lookup(tmplKey)`. Based on the rest of the stack trace and resulting panic, `rd.target.templates()` is returing nil, which then makes Lookup panic a little further down the stack.

Working back from that, `t.rootTemplate` in backend/task.go is what `rd.target.templates()` is likely returning here, since the crashing request was for actions related to a job for task krib-install.

So, where does rootTemplate get populated? In Task.Validate, conditional on t.Useable() being true, which seems to take us over to `models/validate.go`, where `v.Validated` is what we're checking. That is set in task.go by a call to `SetValid()`, which makes `Validated` true iff no errors were recorded.

Combing through various functions in model/, notably MergeTemplates, I see no way for it to return a nil set of templates without marking an error.

And yet, if I `drpcli templates show krib-install.sh.tmpl`, it's `Validated=true`, `Errors=[]`.

So, how do we end up with a nil pointer deref?

**Comment by galthaus** [comment]
3.8.2 - it was in one of the other issues.

**Comment by galthaus** [comment]
I tried this with two machines on packet with centos 7 and tip DRP.  I didn't hit this issue.

I'm going to try with 3.8.2.

**Comment by galthaus** [comment]
Okay - this is indirectly fixed in either tip DRP or tip krib content package.
I suspect that if you look at the tasks for krib-install and krib-config you will find that they are marked invalid.  The system doesn't like their task names.  We fixed that by loosening the name validator on tasks and making the names okay for krib in the content package.  Updating to either will fix the issue in the sense that the failure will go away.

I'm going to take this bug to try and fix up the error path so that the system doesn't panic/recover, but instead logs a message saying content has become invalid.

**Comment by galthaus** [comment]
This is a general issue to increase the error logs around invalid objects that are attempted to be rendered.  The looping panic is expected, but really should be a looping error message.

The challenge is around line 147 in backend/renderData.go

**Comment by svallebro** [comment]
I might have come across the same issue.
When looking in the UI on the task krib-install and krib-config. 
The UI complains about the name of the task.

I tried to clone and give it another name and the UI did not complain anymore.
The problem seems to be the ( ) in the "Config Kubernetes (kubeadm)".

**Comment by sygibson** [comment]
@svallebro - the current `stable` version still exhibits this problem - I believe that our `tip` version has a fix in it already to address the issue.  If you get the opportunity, could you please test with the latest `tip` version and verify if the problem you are seeing still exists? 

If you're running an isolated mode, `sudo pkill dr-provision`, then do:

`curl get.rebar.digital/tip | bash -s -- upgrade --isolated` 

*from* the directory you did the original install in.  If it is not an isolated mode, simply drop the `--isolated` flag.   

You will also want to go to the Contents menu panel and update the Community Content and the KRIB content packs as well.  

Thank you.

**Comment by zehicle** [comment]
Fixed in 3.10 release.  Case is caught and thrown as a log message.  Will also put an error on the machine.
